### ISSUE TITLE
L3 worker cannot see host tensors created after worker startup

### ISSUE THREAD

**Issue by ndleslx** [ChaoWao/simpler#1027]
## Summary

`DistributedWorker` / L3 worker currently requires all host input/output tensors passed to `worker.run(...)` to be created and moved to shared memory before the L3 worker child process is created. If a later run uses a newly-created tensor, or a tensor with a shape/storage that was not shared before worker creation, the child process cannot reliably see that storage.

This creates a practical limitation for dynamic-shape serving workloads: every possible input/output buffer shape must be preallocated up front, otherwise repeated `worker.run(...)` calls with changed input shapes may fail or require awkward fixed-size buffer workarounds.

## Context

This was observed while migrating `pypto-serving` Qwen3-14B prefill/decode dispatch from the L2 worker path to a shared L3 `DistributedWorker` path:

- Feature issue: https://github.com/hw-native-sys/pypto-serving/issues/26
- Implementation PR: https://github.com/hw-native-sys/pypto-serving/pull/29

In that PR, serving currently has to preallocate fixed-size host buffers for prefill/decode before creating the L3 worker:

- prefill hidden / seq_lens / chunk_lens / chunk_offsets / block_table / slot_mapping / logits
- decode hidden / seq_lens / block_table / slot_mapping / logits

The worker is then created once and reused for both prefill and decode. This works, but only because all runtime input/output buffers have stable shared-memory storage before the child process starts.

## Problem

For dynamic serving workloads, input shapes can change between runs:

- prefill total token count changes per request batch / chunk
- user batch size changes
- block table and slot mapping sizes may vary with max sequence / page layout
- output logits buffers may vary with batch shape in other models

With the current L3 worker behavior, tensors created after the worker child exists are not visible to that child unless they were already backed by shared memory visible at fork/spawn time. This means a natural pattern like:

```python
worker = DistributedWorker([program])

for request_batch in batches:
    hidden = torch.empty((dynamic_tokens, hidden_size), dtype=torch.bfloat16)
    out = torch.empty((dynamic_batch, vocab), dtype=torch.float32)
    worker.run(program, hidden, out)
```

is not safe unless `hidden` and `out` are replaced with slices of pre-created shared-memory buffers.

## Current workaround

`pypto-serving` now preallocates maximum-capacity buffers before L3 worker creation and reuses slices for each run. For example:

- prefill uses a fixed `kernel_batch * max_seq` hidden buffer but passes only the active token slice
- fixed-size metadata buffers are zeroed / filled each run
- decode uses fixed `kernel_batch` buffers and pads inactive rows

This workaround is functional, but it has drawbacks:

- higher memory footprint because maximum-size buffers must exist up front
- model runner code becomes more complex
- dynamic-shape programs need fixed-capacity host buffers even when only a small shape is active
- shape changes are constrained by preallocated capacity
- child-process visibility details leak into serving-side code

## Expected behavior

The L3 worker runtime should support one of the following:

1. Passing newly-created shared-memory host tensors to an already-running L3 worker, with the child process able to attach/map them for that run.
2. A documented runtime API to register or refresh host tensor storage mappings after worker creation.
3. A clear error if a non-visible host tensor is passed to `worker.run(...)`, instead of failing indirectly or forcing users to discover the preallocation requirement.
4. Documentation explaining the exact lifetime/visibility rules for host tensor arguments used by `DistributedWorker.run(...)`.

## Why this matters

Serving engines naturally have dynamic request batches. Requiring every possible host input/output tensor to be created before worker startup makes the L3 worker path harder to use as a general replacement for the L2 runtime path.

The limitation is manageable for Qwen3-14B today through fixed-size buffers, but it becomes fragile for other models, variable vocab/logit outputs, dynamic prefill chunking, or future serving paths that need more flexible input shapes.

## Validation from pypto-serving

After applying the fixed preallocation workaround in `pypto-serving`, these checks pass:

- offline generation without larger PTO2 runtime envs
- single-request HTTP serving without larger PTO2 runtime envs
- multi-request HTTP serving with:

```text
PTO2_RING_HEAP=4294967296
PTO2_RING_TASK_WINDOW=1048576
PTO2_RING_DEP_POOL=1048576
```

This confirms the workaround is viable, but the underlying L3 child-process tensor visibility limitation remains.

**Comment by ChaoZheng109** [comment]
## Root cause

The L3 worker (`Worker(level=3)`) forks its chip children via `os.fork()` (worker.py:1951/1971/2018). At `run()` time the orch fn runs in the **parent** and issues `orch.copy_to(chip, dst, src=tensor.data_ptr(), size)`; the child reads that **raw virtual address** directly via `cw.copy_to(dst, src, n)` (worker.py:861-865).

So a host tensor is only usable if it was `.share_memory_()`-mapped (`MAP_SHARED`) **before** the fork — only then does it live at the *same* virtual address in both parent and child. A tensor created **after** fork (even with `.share_memory_()`) is either unmapped in the child or maps to unrelated memory → crash or silent garbage. That is exactly the "preallocate every buffer before worker creation" constraint this issue describes.

## Good news: the mechanism already exists

`_CTRL_REGISTER` (worker.py:879-926) already solves the structurally identical problem for **callables** registered after fork: the parent stages bytes into a freshly-created POSIX `SharedMemory`, passes the `shm_name` through the mailbox, and the child `SharedMemory(name=...)` mmaps it **on demand**. This "pass a shm by name, child maps it post-fork" path is what host tensors should reuse.

## Decision

We will ship **B + C + D** (explicit registration, not the fully-transparent variant):

- **B — explicit host-buffer registration API.** Add `Worker.register_host_buffer(tensor)` / `unregister_host_buffer(...)`, backed by new mailbox sub-commands (`_CTRL_MAP_HOST` / `_CTRL_UNMAP_HOST`) that mirror `_CTRL_REGISTER`. The child maps the buffer by name once and reuses it. Addresses expectation **#2**.
- **C — clear error on a non-visible tensor.** Passing a host tensor that is neither fork-time-inherited nor registered raises an actionable error pointing the caller at `register_host_buffer`, instead of failing indirectly. Addresses expectation **#3**. Most of C falls out of B's rejection path for free.
- **D — document the lifetime/visibility contract** in `docs/comm-domain.md` (extending the existing `.share_memory_()`-before-`init()` note), addressing expectation **#4**.

### Why B over the fully-transparent option (A)

A transparent variant ("pass any newly-created tensor, the runtime auto-maps it") was considered and deferred. It would have to reverse-map a raw `data_ptr` back to "is this visible in the child?" via an address-interval table, plus a cache with weakref-driven invalidation. A torch tensor can be GC'd and its `data_ptr` immediately reused by a new tensor, so a stale cached mapping → **silent memory corruption**, the worst failure class in a forked runtime. B keys off the buffer the user explicitly hands over, so it avoids address-guessing entirely and gives C for free.

The two are not mutually exclusive and are naturally layered: **B is the mechanism, A is sugar**. If full transparency is wanted later, it can be added on top of B (auto-register on first sight + weakref auto-unregister) without taking on that risk up front. For serving workloads that already manage a buffer pool, "register once, reuse" is sufficient and removes the max-size preallocation workaround.

## Implementation notes

- **Write-back / zero-copy is the real risk, independent of B vs A.** First version will use a runtime-owned named shm + `memcpy` (simple, correctness-first); true zero-copy (mapping the tensor's own storage, which needs SCM_RIGHTS fd-passing or torch's `file_system` sharing strategy) is left as a follow-up optimization.
- **No new env/macro gate.** The new mailbox sub-commands are a protocol extension, not a behavior gate; the capability is unconditional.
- A failing repro test (fork → new tensor → child reads garbage/crashes) lands first and is what B turns green; it also exercises C's error path.
